c10/util/BFloat16.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+// Defines the bloat16 type (brain floating-point). This representation uses
+// 1 bit for the sign, 8 bits for the exponent and 7 bits for the mantissa.
+
+#include <c10/macros/Macros.h>
+#include <cmath>
+#include <cstring>
+
+namespace c10 {
+
+namespace detail {
+  inline C10_HOST_DEVICE float f32_from_bits(uint16_t src) {
+    float res = 0;
+    uint32_t tmp = src;
+    tmp <<= 16;
+
+#ifdef __HIP_PLATFORM_HCC__
+    float* tempRes = 0;
+
+    // We should be using memcpy in order to respect the strict aliasing rule
+    // but it fails in the HIP environment.
+    tempRes = reinterpret_cast<float*>(&tmp);
+    res = *tempRes;
+#else
+    std::memcpy(&res, &tmp, sizeof(tmp));
```

Comment:
If I remember correctly, the CUDA compiler doesn't optimize this memcpy well. I believe you use this in device code in your subsequent PR. You may want to check the CUDA performance and/or use the HIP codepath for CUDA as well.

Reply:
I think it's fine now, in cuda 9.2 at least https://godbolt.org/z/024phL